### ISSUE TITLE
How to improve my training data?

### ISSUE THREAD

**Issue by Karol-G** [AlexeyAB/darknet#3291]
Hi,

I am relative new to the field of neural networks.
I want to detect game figures in a board game called "Risiko" and created a small dataset for that reason.
There are three classes in total: Infantry, horse, and cannon (The color of a figure does not matter).
Because creating the training data takes quite a while, I put 112 figures on each photo (7 figures of one class and one color * 3 classes * 4 different colors -> Everything should be perfectly balanced)
I then took four photos on similar top-down-angles of the scene.
In total I did this 5 times in different rooms with different illuminations. So I got 20 images with a total of 560 objects per class (I know 2000 objects per class are recommended).
Using map during training I achieved an accuracy of 99% on the test set.
However, when I tested with new unknown data, my detection failed miserably.

So my questions are:

If I simply create much more training data, will this solve my problem?

When testing with new unknown data, the illumination was different. Did my model overfit to the 5 different illuminations of my training data? Doesn't yolo do data augmentation for this reason?

Does it make a difference if I have 20 images with 560 objects of one class or 560 images with one object of that class?

When creating the training data my figures are mostly spread out, but in reality they can be quite close to each other.
Does this have an impact of detecting figures close to each other or is it not relevant how the figures are located in my training data?

My cfg:

<details><summary>**Click to open**</summary>
<p>

[net]
batch=64
subdivisions=32
width=608
height=608
channels=3
momentum=0.9
decay=0.0005
angle=0
saturation = 1.5
exposure = 1.5
hue=.1

learning_rate=0.001
burn_in=400
max_batches=6000
policy=steps
steps=5400
scales=.1

[convolutional]
batch_normalize=1
filters=32
size=3
stride=1
pad=1
activation=leaky

//Downsample

[convolutional]
batch_normalize=1
filters=64
size=3
stride=2
pad=1
activation=leaky

[convolutional]
batch_normalize=1
filters=32
size=1
stride=1
pad=1
activation=leaky

[convolutional]
batch_normalize=1
filters=64
size=3
stride=1
pad=1
activation=leaky

[shortcut]
from=-3
activation=linear

// Downsample

[convolutional]
batch_normalize=1
filters=128
size=3
stride=2
pad=1
activation=leaky

[convolutional]
batch_normalize=1
filters=64
size=1
stride=1
pad=1
activation=leaky

[convolutional]
batch_normalize=1
filters=128
size=3
stride=1
pad=1
activation=leaky

[shortcut]
from=-3
activation=linear

[convolutional]
batch_normalize=1
filters=64
size=1
stride=1
pad=1
activation=leaky

[convolutional]
batch_normalize=1
filters=128
size=3
stride=1
pad=1
activation=leaky

[shortcut]
from=-3
activation=linear

// Downsample

[convolutional]
batch_normalize=1
filters=256
size=3
stride=2
pad=1
activation=leaky

[convolutional]
batch_normalize=1
filters=128
size=1
stride=1
pad=1
activation=leaky

[convolutional]
batch_normalize=1
filters=256
size=3
stride=1
pad=1
activation=leaky

[shortcut]
from=-3
activation=linear

[convolutional]
batch_normalize=1
filters=128
size=1
stride=1
pad=1
activation=leaky

[convolutional]
batch_normalize=1
filters=256
size=3
stride=1
pad=1
activation=leaky

[shortcut]
from=-3
activation=linear

[convolutional]
batch_normalize=1
filters=128
size=1
stride=1
pad=1
activation=leaky

[convolutional]
batch_normalize=1
filters=256
size=3
stride=1
pad=1
activation=leaky

[shortcut]
from=-3
activation=linear

[convolutional]
batch_normalize=1
filters=128
size=1
stride=1
pad=1
activation=leaky

[convolutional]
batch_normalize=1
filters=256
size=3
stride=1
pad=1
activation=leaky

[shortcut]
from=-3
activation=linear

[convolutional]
batch_normalize=1
filters=128
size=1
stride=1
pad=1
activation=leaky

[convolutional]
batch_normalize=1
filters=256
size=3
stride=1
pad=1
activation=leaky

[shortcut]
from=-3
activation=linear

[convolutional]
batch_normalize=1
filters=128
size=1
stride=1
pad=1
activation=leaky

[convolutional]
batch_normalize=1
filters=256
size=3
stride=1
pad=1
activation=leaky

[shortcut]
from=-3
activation=linear

[convolutional]
batch_normalize=1
filters=128
size=1
stride=1
pad=1
activation=leaky

[convolutional]
batch_normalize=1
filters=256
size=3
stride=1
pad=1
activation=leaky

[shortcut]
from=-3
activation=linear

[convolutional]
batch_normalize=1
filters=128
size=1
stride=1
pad=1
activation=leaky

[convolutional]
batch_normalize=1
filters=256
size=3
stride=1
pad=1
activation=leaky

[shortcut]
from=-3
activation=linear

// Downsample

[convolutional]
batch_normalize=1
filters=512
size=3
stride=2
pad=1
activation=leaky

[convolutional]
batch_normalize=1
filters=256
size=1
stride=1
pad=1
activation=leaky

[convolutional]
batch_normalize=1
filters=512
size=3
stride=1
pad=1
activation=leaky

[shortcut]
from=-3
activation=linear

[convolutional]
batch_normalize=1
filters=256
size=1
stride=1
pad=1
activation=leaky

[convolutional]
batch_normalize=1
filters=512
size=3
stride=1
pad=1
activation=leaky

[shortcut]
from=-3
activation=linear

[convolutional]
batch_normalize=1
filters=256
size=1
stride=1
pad=1
activation=leaky

[convolutional]
batch_normalize=1
filters=512
size=3
stride=1
pad=1
activation=leaky

[shortcut]
from=-3
activation=linear

[convolutional]
batch_normalize=1
filters=256
size=1
stride=1
pad=1
activation=leaky

[convolutional]
batch_normalize=1
filters=512
size=3
stride=1
pad=1
activation=leaky

[shortcut]
from=-3
activation=linear

[convolutional]
batch_normalize=1
filters=256
size=1
stride=1
pad=1
activation=leaky

[convolutional]
batch_normalize=1
filters=512
size=3
stride=1
pad=1
activation=leaky

[shortcut]
from=-3
activation=linear

[convolutional]
batch_normalize=1
filters=256
size=1
stride=1
pad=1
activation=leaky

[convolutional]
batch_normalize=1
filters=512
size=3
stride=1
pad=1
activation=leaky

[shortcut]
from=-3
activation=linear

[convolutional]
batch_normalize=1
filters=256
size=1
stride=1
pad=1
activation=leaky

[convolutional]
batch_normalize=1
filters=512
size=3
stride=1
pad=1
activation=leaky

[shortcut]
from=-3
activation=linear

[convolutional]
batch_normalize=1
filters=256
size=1
stride=1
pad=1
activation=leaky

[convolutional]
batch_normalize=1
filters=512
size=3
stride=1
pad=1
activation=leaky

[shortcut]
from=-3
activation=linear

// Downsample

[convolutional]
batch_normalize=1
filters=1024
size=3
stride=2
pad=1
activation=leaky

[convolutional]
batch_normalize=1
filters=512
size=1
stride=1
pad=1
activation=leaky

[convolutional]
batch_normalize=1
filters=1024
size=3
stride=1
pad=1
activation=leaky

[shortcut]
from=-3
activation=linear

[convolutional]
batch_normalize=1
filters=512
size=1
stride=1
pad=1
activation=leaky

[convolutional]
batch_normalize=1
filters=1024
size=3
stride=1
pad=1
activation=leaky

[shortcut]
from=-3
activation=linear

[convolutional]
batch_normalize=1
filters=512
size=1
stride=1
pad=1
activation=leaky

[convolutional]
batch_normalize=1
filters=1024
size=3
stride=1
pad=1
activation=leaky

[shortcut]
from=-3
activation=linear

[convolutional]
batch_normalize=1
filters=512
size=1
stride=1
pad=1
activation=leaky

[convolutional]
batch_normalize=1
filters=1024
size=3
stride=1
pad=1
activation=leaky

[shortcut]
from=-3
activation=linear

////////////////////////////////////////////

[convolutional]
batch_normalize=1
filters=512
size=1
stride=1
pad=1
activation=leaky

[convolutional]
batch_normalize=1
size=3
stride=1
pad=1
filters=1024
activation=leaky

[convolutional]
batch_normalize=1
filters=512
size=1
stride=1
pad=1
activation=leaky

[convolutional]
batch_normalize=1
size=3
stride=1
pad=1
filters=1024
activation=leaky

[convolutional]
batch_normalize=1
filters=512
size=1
stride=1
pad=1
activation=leaky

[convolutional]
batch_normalize=1
size=3
stride=1
pad=1
filters=1024
activation=leaky

[convolutional]
size=1
stride=1
pad=1
// filters = (num/3) * (5+classes)
filters=24
activation=linear

[yolo]
mask = 6,7,8
anchors = 10,13,  16,30,  33,23,  30,61,  62,45,  59,119,  116,90,  156,198,  373,326
classes=3
num=9
jitter=.3
ignore_thresh = .5
truth_thresh = 1
random=1

[route]
layers = -4

[convolutional]
batch_normalize=1
filters=256
size=1
stride=1
pad=1
activation=leaky

[upsample]
stride=2

[route]
layers = -1, 61

[convolutional]
batch_normalize=1
filters=256
size=1
stride=1
pad=1
activation=leaky

[convolutional]
batch_normalize=1
size=3
stride=1
pad=1
filters=512
activation=leaky

[convolutional]
batch_normalize=1
filters=256
size=1
stride=1
pad=1
activation=leaky

[convolutional]
batch_normalize=1
size=3
stride=1
pad=1
filters=512
activation=leaky

[convolutional]
batch_normalize=1
filters=256
size=1
stride=1
pad=1
activation=leaky

[convolutional]
batch_normalize=1
size=3
stride=1
pad=1
filters=512
activation=leaky

[convolutional]
size=1
stride=1
pad=1
// filters = (num/3) * (5+classes)
filters=24
activation=linear

[yolo]
mask = 3,4,5
anchors = 10,13,  16,30,  33,23,  30,61,  62,45,  59,119,  116,90,  156,198,  373,326
classes=3
num=9
jitter=.3
ignore_thresh = .5
truth_thresh = 1
random=1

[route]
layers = -4

[convolutional]
batch_normalize=1
filters=128
size=1
stride=1
pad=1
activation=leaky

[upsample]
stride=4

[route]
layers = -1, 11

[convolutional]
batch_normalize=1
filters=128
size=1
stride=1
pad=1
activation=leaky

[convolutional]
batch_normalize=1
size=3
stride=1
pad=1
filters=256
activation=leaky

[convolutional]
batch_normalize=1
filters=128
size=1
stride=1
pad=1
activation=leaky

[convolutional]
batch_normalize=1
size=3
stride=1
pad=1
filters=256
activation=leaky

[convolutional]
batch_normalize=1
filters=128
size=1
stride=1
pad=1
activation=leaky

[convolutional]
batch_normalize=1
size=3
stride=1
pad=1
filters=256
activation=leaky

[convolutional]
size=1
stride=1
pad=1
// filters = (num/3) * (5+classes)
filters=24
activation=linear

[yolo]
mask = 0,1,2
anchors = 10,13,  16,30,  33,23,  30,61,  62,45,  59,119,  116,90,  156,198,  373,326
classes=3
num=9
jitter=.3
ignore_thresh = .5
truth_thresh = 1
random=1

</p>
</details>

Some examples of my training data:

![1](https://user-images.githubusercontent.com/3471895/58702923-6a6fce00-83a7-11e9-8b85-fe98c7c75e49.jpg)
![2](https://user-images.githubusercontent.com/3471895/58702946-75c2f980-83a7-11e9-9465-0d340660a7b9.jpg)

**Comment by AlexeyAB** [comment]
@CookingCookie Hi,

> However, when I tested with new unknown data, my detection failed miserably.

Can you show examples of "new unknown data"?

Can you show some labeled trainin images, which the most similar to the "new unknown data"?

**Comment by CookingCookie** [comment]
Labeled training image (probably most similar to new unknown data):

![Unbenannt](https://user-images.githubusercontent.com/3471895/58705119-81fe8500-83ae-11e9-9c07-b0e6126f63db.PNG)


Detected objects on one of the images in test set:
(Blue box = infantry, green = horse, red = cannon)

![resultImage28](https://user-images.githubusercontent.com/3471895/58704966-0ac8f100-83ae-11e9-81d8-8622ba7d64fa.jpg)


Detected objects on new unknown data (different room and day time -> different illumination):
(Blue box = infantry, green = horse, red = cannon)

![resultImage6](https://user-images.githubusercontent.com/3471895/58705007-351aae80-83ae-11e9-915a-1d75dff07b7d.jpg)

In this image some objects are classified incorrect and the cluster of blue infantry in the center is not detected as single units.

**Comment by AlexeyAB** [comment]
How many iterations did you train?

Yolo can augment colors, scales and aspect ratio, but can't augment tilt.

So add the images from different angles and tilt to the training dataset.

And you can try train by using this model https://github.com/AlexeyAB/darknet/files/3199657/yolo_v3_spp_pan.cfg.txt
Just set `max=200` in the last `[yolo]` layer.

Also you can try to train with 
```
width=608
height=608
saturation = 1.8
exposure = 1.8
hue=.3
```
so the model will be more resistant to changing lighting.

**Comment by CookingCookie** [comment]
> How many iterations did you train?

First I trained with net size 416x416 for 6000 iterations. I then tested with the best weights and the weights at iteration 1000, 2000 and so on. Currently I am training with net size 608x608 at iteration 800 and tested so far with iteration 400 and 800.

> Yolo can augment colors, scales and aspect ratio, but can't augment tilt.
> 
> So add the images from different angles and tilt to the training dataset.
> 
> And you can try train by using this model > > https://github.com/AlexeyAB/darknet/files/3199657/yolo_v3_spp_pan.cfg.txt
> Just set max=200 in the last [yolo] layer.
> 
> Also you can try to train with
> 
> width=608
> height=608
> saturation = 1.8
> exposure = 1.8
> hue=.3
> so the model will be more resistant to changing lighting.

Thanks I will try this!

If it doesn't bother you, I would still like to know what you think about this:

Does it make a difference if I have 20 images with 560 objects of one class or 560 images with one object of that class?

When creating the training data my figures are mostly spread out, but in reality they can be quite close to each other.
Does this have an impact of detecting figures close to each other or is it not relevant how the figures are located in my training data?

**Comment by CookingCookie** [comment]
When using the yolo_v3_spp_pan.cfg I get the following error:
`Error: l.outputs == params.inputs 
filters= in the [convolutional]-layer doesn't correspond to classes= or mask= in [yolo]-layer `

I only changed the number of classes to 3.

**Comment by AlexeyAB** [comment]
> I only changed the number of classes to 3.

You also should change filters, look that num=12 not 9 in new cfg-file
```
// filters = (num/3) * (5+classes)
filters=32
```

>  Currently I am training with net size 608x608 at iteration 800 and tested with iteration 400 and 800 so far.

You shouldn't test it at iteration less than 1000.


> Does it make a difference if I have 20 images with 560 objects of one class or 560 images with one object of that class?

If on your test images will be 560 objects on one image, then you should use the same training images - 560 objects on one image. 
Just you should set `max=560` in the last yolo-layer.

> Does this have an impact of detecting figures close to each other or is it not relevant how the figures are located in my training data?

You should allocate your figures in the same way, as they will be in the test images.

**Comment by DoriHp** [comment]
@AlexeyAB can I change width, height and archors during training process to improve accuracy?
